src/database/catalog.cpp

```diff
@@ -1,69 +1,90 @@
+  p_fields = new std::unordered_map<int, std::string>();
 }
 
 void Catalog::AddTable(DbFile* file, std::string name, std::string pkeyField) {
-  Table* tab = new Table(file, name, pkeyField);
-  TableList.push_back(file);
-}
+  int table_id = file->get_id();
 
-int Catalog::get_table_id(std::string name) {
-  for (int i=0; i<TableList.size(); i++) {
-    if (TableList[i].name == name) {
-      return TableList[i].file->get_id();
+  for (auto const& elem: names) {
+    int id = elem.first;
+    if (names[id] == name) {
+      if (id != table_id) {
+        names.erase(id);
+        db_fields.erase(id);
+        p_fields.erase(id);
+      }
+      break;
     }
   }
-  // throw new NoSuchElementException();
+  names[table_id] = name;
+  db_fields[table_id] = file;
+  p_fields[table_id] = pkeyField;
 }
 
-std::string Catalog::get_table_name(int table_id) {
-  for (int i=0; i<TableList.size(); i++) {
-    if (TableList[i].file->get_id() == table_id) {
-      return TableList[i].name;
+void Catalog::AddTable(DbFile* file, std::string name) {
+  AddTable(file, name, "");
+}
+
+void Catalog::AddTable(DbFile* file) {
+  // addTable(file, (UUID.randomUUID()).toString());
+}
+
+int Catalog::get_table_id(std::string name) {
+  for (auto const& elem: names) {
+    int id = elem.first;
+    if (name[id] == name) {
+      return id;
     }
   }
-
   // throw new NoSuchElementException();
 }
-
 TupleDesc * Catalog::get_tuple_desc(int table_id) {
-  for (int i = 0; i < TableList.size(); i++) {
-    if (TableList[i].file->get_id() == table_id) {
-      return TableList[i].file->getTupleDesc();
-    }
+  if (db_fields.count(table_id)) {
+    return db_fields[table_id]->get_tuple_desc();
   }
   // throw new NoSuchElementException();
 }
 
-DbFile * get_db_file(int table_id) {
+std::string Catalog::get_table_name(int table_id) {
+  if (names.count(table_id)) {
+    return names[table_id];
+  }
+  // throw new NoSuchElementException();
 }
 
-std::string Catalog::GetPrimaryKey(int tableid) {
-  for (int i=0; i<TableList.size(); i++) {
-    if (TableList[i].file->get_id() == tableid) {
-      return TableList[i].file->pkeyField;
-    }
+DbFile * get_db_file(int table_id) {
+  if (db_fields.count(table_id)) {
+    return db_fields[table_id];
+  }
+  for (auto const& elem: db_fields) {
+    int id = elem.first;
+    std::cout << "has:\t " << id << " " << names[id] << std::endl;
   }
+  std::cout << table_id << std::endl;
   // throw new NoSuchElementException();
 }
 
-Iterator<Integer> Catalog::TableIdIterator() {
-  std::vector<int> lst;
-  for (Table* tab: TableList) {
-    lst.push_back(tab.file->get_id());
+std::string Catalog::GetPrimaryKey(int table_id) {
+  if (p_fields.count(table_id)) {
+    return p_fields[table_id];
   }
+  // throw new NoSuchElementException();
+}
 
-  // Iterator<Integer> it = lst.iterator();
-  // return it
+Iterator Catalog::TableIdIterator() {
+  std::unordered_map<int, DbFile*>::iter = db_fields.begin();
```

Comment:
Really glad to see you figured this out :) great job Benji!

The function signature is wrong, thanks to me, but I'll fix it after we merge!